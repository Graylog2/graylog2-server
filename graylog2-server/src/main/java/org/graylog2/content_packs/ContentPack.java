/**
 * This file is part of Graylog.
 *
 * Graylog is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Graylog is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Graylog.  If not, see <http://www.gnu.org/licenses/>.
 */
package org.graylog2.content_packs;

import com.fasterxml.jackson.annotation.JsonAutoDetect;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonProperty;
import org.mongojack.Id;
import org.mongojack.ObjectId;

import javax.validation.constraints.NotBlank;
import javax.validation.constraints.NotNull;
import java.util.Collections;
import java.util.Set;

@JsonAutoDetect
@JsonInclude(JsonInclude.Include.NON_NULL)
public class ContentPack {
    @Id
    @ObjectId
    @JsonProperty
    @NotBlank
    private String id;
    @JsonProperty
    @NotBlank
    private String name;
    @JsonProperty
    @NotBlank
    private String version;
    @JsonProperty
    private String description;
    @JsonProperty
    private String summary;
    @JsonProperty
    @NotNull
    private Set<String> states = Collections.emptySet();

    public String getId() {
        return id;
    }

    public void setId(String id) {
        this.id = id;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getVersion() {
        return version;
    }

    public void setVersion(String version) {
        this.version = version;
    }

    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public String getSummary() {
        return summary;
    }

    public void setSummary(String summary) {
        this.summary = summary;
    }

    public Set<String> getStates() {
        return states;
    }

    public void setStates(Set<String> states) {
        this.states = states;
    }
}
