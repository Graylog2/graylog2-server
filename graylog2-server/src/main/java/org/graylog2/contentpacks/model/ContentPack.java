/**
 * This file is part of Graylog.
 *
 * Graylog is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Graylog is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Graylog.  If not, see <http://www.gnu.org/licenses/>.
 */
package org.graylog2.contentpacks.model;

import com.fasterxml.jackson.annotation.JsonSubTypes;
import com.fasterxml.jackson.annotation.JsonTypeInfo;
import org.graylog2.contentpacks.model.constraints.Constraint;

import java.util.Set;

@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, include = JsonTypeInfo.As.EXISTING_PROPERTY, property = Versioned.FIELD_META_VERSION, defaultImpl = LegacyContentPack.class)
@JsonSubTypes({
        @JsonSubTypes.Type(value = LegacyContentPack.class),
        @JsonSubTypes.Type(value = ContentPackV1.class, name = ContentPackV1.VERSION)
})
public interface ContentPack extends Identified, Revisioned, Versioned {
    interface ContentPackBuilder<SELF> extends IdBuilder<SELF>, RevisionBuilder<SELF>, VersionBuilder<SELF> {
    }
}
