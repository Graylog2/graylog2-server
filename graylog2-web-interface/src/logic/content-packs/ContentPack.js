import { Map, Set } from 'immutable';
import { concat, remove } from 'lodash';
import uuid from 'uuid/v4';
import Entity from './Entity';

export default class ContentPack {
  constructor(v, id, rev, name, summary, description, vendor, url, parameters, entitieValues) {
    const entities = entitieValues.map((e) => {
      if (e instanceof Entity) {
        return e;
      }
      return Entity.fromJSON(e, false, parameters);
    });

    this._value = {
      v,
      id,
      rev,
      name,
      summary,
      description,
      vendor,
      url,
      parameters,
      entities,
    };
  }

  get v() {
    return this._value.v;
  }

  get id() {
    return this._value.id;
  }

  get rev() {
    return this._value.rev;
  }

  get name() {
    return this._value.name;
  }

  get summary() {
    return this._value.summary;
  }

  get description() {
    return this._value.description;
  }

  get vendor() {
    return this._value.vendor;
  }

  get url() {
    return this._value.url;
  }

  get constraints() {
    return this._value.entities.reduce((acc, entity) => {
      return entity.constraints.reduce((result, constraint) => {
        return result.add(constraint);
      }, acc);
    }, Set());
  }

  get parameters() {
    return this._value.parameters;
  }

  get entities() {
    return this._value.entities;
  }

  toBuilder() {
    const {
      v,
      id,
      rev,
      name,
      summary,
      description,
      vendor,
      url,
      parameters,
      entities,
    } = this._value;
    // eslint-disable-next-line no-use-before-define
    return new Builder(Map({
      v,
      id,
      rev,
      name,
      summary,
      description,
      vendor,
      url,
      parameters,
      entities,
    }));
  }

  toJSON() {
    const {
      v,
      id,
      rev,
      name,
      summary,
      description,
      vendor,
      url,
      parameters,
      entities,
    } = this._value;

    const entitiesJSON = entities.map(e => e.toJSON());
    return {
      v,
      id,
      rev,
      name,
      summary,
      description,
      vendor,
      url,
      parameters,
      entities: entitiesJSON,
    };
  }

  static fromJSON(value) {
    const {
      v,
      id,
      rev,
      name,
      summary,
      description,
      vendor,
      url,
      parameters,
      entities,
    } = value;
    return new ContentPack(
      v,
      id,
      rev,
      name,
      summary,
      description,
      vendor,
      url,
      parameters,
      entities,
    );
  }

  static builder() {
    // eslint-disable-next-line no-use-before-define
    return new Builder()
      .v(1)
      .id(uuid())
      .rev(1)
      .name('')
      .summary('')
      .description('')
      .vendor('')
      .url('')
      .parameters([])
      .entities([]);
  }
}

class Builder {
  constructor(value = Map()) {
    this.value = value;
  }

  v(value) {
    this.value = this.value.set('v', value);
    return this;
  }

  id(value) {
    this.value = this.value.set('id', value);
    return this;
  }

  rev(value) {
    this.value = this.value.set('rev', value);
    return this;
  }

  name(value) {
    this.value = this.value.set('name', value);
    return this;
  }

  summary(value) {
    this.value = this.value.set('summary', value);
    return this;
  }

  description(value) {
    this.value = this.value.set('description', value);
    return this;
  }

  vendor(value) {
    this.value = this.value.set('vendor', value);
    return this;
  }

  url(value) {
    this.value = this.value.set('url', value);
    return this;
  }

  parameters(value) {
    this.value = this.value.set('parameters', value);
    return this;
  }

  removeParameter(value) {
    const parameters = this.value.get('parameters').slice(0);
    remove(parameters, parameter => parameter.name === value.name);
    this.value = this.value.set('parameters', parameters);
    return this;
  }

  addParameter(value) {
    const parameters = this.value.get('parameters');
    const newParameters = concat(parameters, value);
    this.value = this.value.set('parameters', newParameters);
    return this;
  }

  entities(value) {
    this.value = this.value.set('entities', value);
    return this;
  }

  build() {
    const {
      v,
      id,
      rev,
      name,
      summary,
      description,
      vendor,
      url,
      parameters,
      entities,
    } = this.value.toObject();
    return new ContentPack(
      v,
      id,
      rev,
      name,
      summary,
      description,
      vendor,
      url,
      parameters,
      entities,
    );
  }
}
