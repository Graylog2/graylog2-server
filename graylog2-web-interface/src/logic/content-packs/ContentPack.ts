/*
 * Copyright (C) 2020 Graylog, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the Server Side Public License, version 1,
 * as published by MongoDB, Inc.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * Server Side Public License for more details.
 *
 * You should have received a copy of the Server Side Public License
 * along with this program. If not, see
 * <http://www.mongodb.com/licensing/server-side-public-license>.
 */
import { Map, Set } from 'immutable';
import concat from 'lodash/concat';
import remove from 'lodash/remove';

import generateId from 'logic/generateId';
import type { ContentPackEntity, ContentPackParameter } from 'components/content-packs/Types';

import Entity from './Entity';

export default class ContentPack {
  private _value: {
    v: number;
    id: string;
    rev: number;
    name: string;
    summary: string;
    description: string;
    vendor: string;
    url: string;
    parameters: ContentPackParameter[];
    entities: ContentPackEntity[];
  };

  constructor(
    v: number,
    id: string,
    rev: number,
    name: string,
    summary: string,
    description: string,
    vendor: string,
    url: string,
    parameters: ContentPackParameter[],
    entitieValues: ContentPackEntity[],
  ) {
    const entities = entitieValues.map((e) => {
      if (e instanceof Entity) {
        return e;
      }

      return Entity.fromJSON(e, false, parameters);
    });

    this._value = {
      v,
      id,
      rev,
      name,
      summary,
      description,
      vendor,
      url,
      parameters,
      entities,
    };
  }

  get v() {
    return this._value.v;
  }

  get id() {
    return this._value.id;
  }

  get rev() {
    return this._value.rev;
  }

  get name() {
    return this._value.name;
  }

  get summary() {
    return this._value.summary;
  }

  get description() {
    return this._value.description;
  }

  get vendor() {
    return this._value.vendor;
  }

  get url() {
    return this._value.url;
  }

  get constraints() {
    return this._value.entities.reduce(
      (acc, entity) => entity.constraints.reduce((result, constraint) => result.add(constraint), acc),
      Set(),
    );
  }

  get parameters() {
    return this._value.parameters;
  }

  get entities() {
    return this._value.entities;
  }

  toBuilder() {
    const { v, id, rev, name, summary, description, vendor, url, parameters, entities } = this._value;

    // eslint-disable-next-line @typescript-eslint/no-use-before-define
    return new Builder(
      Map({
        v,
        id,
        rev,
        name,
        summary,
        description,
        vendor,
        url,
        parameters,
        entities,
      }),
    );
  }

  toJSON() {
    const { v, id, rev, name, summary, description, vendor, url, parameters, entities } = this._value;

    const entitiesJSON = entities.map((e) => e.toJSON());

    return {
      v,
      id,
      rev,
      name,
      summary,
      description,
      vendor,
      url,
      parameters,
      entities: entitiesJSON,
    };
  }

  static fromJSON(value) {
    const { v, id, rev, name, summary, description, vendor, url, parameters, entities } = value;

    return new ContentPack(v, id, rev, name, summary, description, vendor, url, parameters, entities);
  }

  static builder() {
    // eslint-disable-next-line @typescript-eslint/no-use-before-define
    return new Builder()
      .v(1)
      .id(generateId())
      .rev(1)
      .name('')
      .summary('')
      .description('')
      .vendor('')
      .url('')
      .parameters([])
      .entities([]);
  }
}

class Builder {
  private value: Map<string, unknown>;

  constructor(value: Map<string, unknown> = Map()) {
    this.value = value;
  }

  v(value: number) {
    this.value = this.value.set('v', value);

    return this;
  }

  id(value: string) {
    this.value = this.value.set('id', value);

    return this;
  }

  rev(value: number) {
    this.value = this.value.set('rev', value);

    return this;
  }

  name(value: string) {
    this.value = this.value.set('name', value);

    return this;
  }

  summary(value: string) {
    this.value = this.value.set('summary', value);

    return this;
  }

  description(value: string) {
    this.value = this.value.set('description', value);

    return this;
  }

  vendor(value: string) {
    this.value = this.value.set('vendor', value);

    return this;
  }

  url(value: string) {
    this.value = this.value.set('url', value);

    return this;
  }

  parameters(value: ContentPackParameter[]) {
    this.value = this.value.set('parameters', value);

    return this;
  }

  private getParameters() {
    return this.value.get('parameters') as ContentPackParameter[];
  }

  removeParameter(value: ContentPackParameter) {
    const parameters = this.getParameters().slice(0);

    remove(parameters, (parameter) => parameter.name === value.name);
    this.value = this.value.set('parameters', parameters);

    return this;
  }

  addParameter(value: ContentPackParameter) {
    const parameters = this.getParameters();
    const newParameters = concat(parameters, value);

    this.value = this.value.set('parameters', newParameters);

    return this;
  }

  entities(value: Array<ContentPackEntity>) {
    this.value = this.value.set('entities', value);

    return this;
  }

  build() {
    const { v, id, rev, name, summary, description, vendor, url, parameters, entities } = this.value.toObject();

    // @ts-ignore
    return new ContentPack(v, id, rev, name, summary, description, vendor, url, parameters, entities);
  }
}
