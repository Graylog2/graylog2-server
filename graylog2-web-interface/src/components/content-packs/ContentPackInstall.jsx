import PropTypes from 'prop-types';
import React from 'react';

import ContentPack from 'logic/content-packs/ContentPack';

import { Row, Col } from 'react-bootstrap';
import { Input } from 'components/bootstrap';
import ValueRefHelper from 'util/ValueRefHelper';
import ContentPackUtils from './ContentPackUtils';

import ContentPackEntitiesList from './ContentPackEntitiesList';

class ContentPackInstall extends React.Component {
  static propTypes = {
    contentPack: PropTypes.object.isRequired,
    onInstall: PropTypes.func,
  };

  static defaultProps = {
    onInstall: () => {},
  };

  constructor(props) {
    super(props);

    const parameterInput = props.contentPack.parameters.reduce((result, parameter) => {
      if (parameter.default_value) {
        const newResult = result;
        newResult[parameter.name] = ContentPackUtils.convertToString(parameter);
        return newResult;
      }
      return result;
    }, {});

    this.state = {
      parameterInput: parameterInput,
      comment: '',
      errorMessages: {},
    };
  }

  onInstall = () => {
    if (this._validateInput()) {
      const contentPackId = this.props.contentPack.id;
      const contentPackRev = this.props.contentPack.rev;
      const parameters = this._convertedParameters();
      this.props.onInstall(contentPackId, contentPackRev,
        { parameters: parameters, comment: this.state.comment });
    }
  };

  _convertedParameters = () => {
    return Object.keys(this.state.parameterInput).reduce((result, paramName) => {
      const newResult = result;
      const paramType = this.props.contentPack.parameters.find(parameter => parameter.name === paramName).type;
      const value = ContentPackUtils.convertValue(paramType, this.state.parameterInput[paramName]);
      newResult[paramName] = ValueRefHelper.createValueRef(paramType, value);
      return newResult;
    }, {});
  };

  _getValue = (name, value) => {
    const newParameterInput = this.state.parameterInput;
    newParameterInput[name] = value;
    this.setState({ parameterInput: newParameterInput });
  };

  _getComment = (e) => {
    this.setState({ comment: e.target.value });
  };

  _validateInput = () => {
    const { parameterInput } = this.state;
    const errors = this.props.contentPack.parameters.reduce((result, parameter) => {
      if (parameterInput[parameter.name] && parameterInput[parameter.name].length > 0) {
        return result;
      }
      const newResult = result;
      newResult[parameter.name] = 'Needs to be filled.';
      return newResult;
    }, {});
    this.setState({ errorMessages: errors });
    return Object.keys(errors).length <= 0;
  };

  renderParameter(parameter) {
    const error = this.state.errorMessages[parameter.name];
    return (
      <Input name={parameter.name}
             id={parameter.name}
             key={parameter.name}
             type="text"
             maxLength={250}
             value={this.state.parameterInput[parameter.name] || ''}
             onChange={(e) => { this._getValue(parameter.name, e.target.value); }}
             labelClassName="col-sm-3"
             wrapperClassName="col-sm-7"
             label={parameter.title}
             help={error || parameter.description}
             bsStyle={error ? 'error' : undefined}
             required />
    );
  }

  render() {
    const parameterInput = this.props.contentPack.parameters.map((parameter) => {
      return this.renderParameter(parameter);
    });
    const contentPack = ContentPack.fromJSON(this.props.contentPack);

    return (<div>
      <Row>
        <Col smOffset={1} sm={10}>
          <h2>Install comment</h2>
          <br />
          <br />
          <Input name="comment"
                 id="comment"
                 type="text"
                 maxLength={512}
                 value={this.state.comment}
                 onChange={this._getComment}
                 labelClassName="col-sm-3"
                 wrapperClassName="col-sm-7"
                 label="Comment"
          />
        </Col>
      </Row>
      {parameterInput.length > 0 &&
      <Row>
        <Col smOffset={1} sm={10}>
          <h2>Configure Parameter</h2>
          <br />
          <br />
          {parameterInput}
        </Col>
      </Row>}
      <Row>
        <Col smOffset={1} sm={10}>
          <ContentPackEntitiesList contentPack={contentPack} readOnly />
        </Col>
      </Row>
    </div>);
  }
}

export default ContentPackInstall;
