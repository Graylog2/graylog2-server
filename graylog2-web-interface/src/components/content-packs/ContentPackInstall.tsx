/*
 * Copyright (C) 2020 Graylog, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the Server Side Public License, version 1,
 * as published by MongoDB, Inc.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * Server Side Public License for more details.
 *
 * You should have received a copy of the Server Side Public License
 * along with this program. If not, see
 * <http://www.mongodb.com/licensing/server-side-public-license>.
 */
import * as React from 'react';
import styled from 'styled-components';

import ContentPack from 'logic/content-packs/ContentPack';
import { Input } from 'components/bootstrap';
import ValueRefHelper from 'util/ValueRefHelper';
import type { ContentPackInstallation, ContentPackParameter } from 'components/content-packs/Types';
import EntityCreateShareFormGroup from 'components/permissions/EntityCreateShareFormGroup';
import type { EntitySharePayload } from 'actions/permissions/EntityShareActions';

import ContentPackUtils from './ContentPackUtils';
import ContentPackEntitiesList from './ContentPackEntitiesList';

type ContentPackInstallProps = {
  contentPack: ContentPackInstallation;
  onInstall?: (id: string, contentPackRev: number, parameters: unknown, shareRequest: EntitySharePayload) => void;
};

type State = {
  parameterInput: { [key: string]: string };
  comment: string;
  errorMessages: { [key: string]: string };
  shareRequest: EntitySharePayload;
};

const EntitiesListContainer = styled.div`
  margin-top: 20px;
`;

class ContentPackInstall extends React.Component<ContentPackInstallProps, State> {
  static defaultProps = {
    onInstall: () => {},
  };

  constructor(props: ContentPackInstallProps) {
    super(props);

    const parameterInput = props.contentPack.parameters.reduce((result, parameter) => {
      if (parameter.default_value) {
        const newResult = result;

        newResult[parameter.name] = ContentPackUtils.convertToString(parameter);

        return newResult;
      }

      return result;
    }, {});

    this.state = {
      parameterInput: parameterInput,
      comment: '',
      errorMessages: {},
      shareRequest: undefined,
    };
  }

  // eslint-disable-next-line react/no-unused-class-component-methods
  onInstall = () => {
    if (this._validateInput()) {
      const contentPackId = this.props.contentPack.id;
      const contentPackRev = this.props.contentPack.rev;
      const parameters = this._convertedParameters();

      this.props.onInstall(
        contentPackId,
        contentPackRev,
        {
          parameters: parameters,
          comment: this.state.comment,
        },
        this.state.shareRequest,
      );
    }
  };

  _convertedParameters = (): { [parameterName: string]: string } =>
    Object.keys(this.state.parameterInput).reduce((result, paramName) => {
      const newResult = result;
      const paramType = this.props.contentPack.parameters.find((parameter) => parameter.name === paramName).type;
      const value = ContentPackUtils.convertValue(paramType, this.state.parameterInput[paramName]);

      newResult[paramName] = ValueRefHelper.createValueRef(paramType, value);

      return newResult;
    }, {});

  _getValue = (name: string, value: string) => {
    this.setState(({ parameterInput }) => ({ parameterInput: { ...parameterInput, [name]: value } }));
  };

  _getComment = (e: React.ChangeEvent<HTMLInputElement>) => {
    this.setState({ comment: e.target.value });
  };

  _validateInput = () => {
    const { parameterInput } = this.state;
    const errors = this.props.contentPack.parameters.reduce((result, parameter) => {
      if (parameterInput[parameter.name]?.length) {
        return result;
      }

      const newResult = result;

      newResult[parameter.name] = 'Needs to be filled.';

      return newResult;
    }, {});

    this.setState({ errorMessages: errors });

    return Object.keys(errors).length <= 0;
  };

  _setShareRequest = (shareRequest: EntitySharePayload) => this.setState({ shareRequest });

  renderParameter(parameter: ContentPackParameter) {
    const error = this.state.errorMessages[parameter.name];

    return (
      <Input
        name={parameter.name}
        id={parameter.name}
        key={parameter.name}
        type="text"
        maxLength={250}
        value={this.state.parameterInput[parameter.name] || ''}
        onChange={(e) => {
          this._getValue(parameter.name, e.target.value);
        }}
        label={parameter.title}
        help={error || parameter.description}
        bsStyle={error ? 'error' : undefined}
        required
      />
    );
  }

  render() {
    const parameterInputs = this.props.contentPack.parameters.map((parameter) => this.renderParameter(parameter));
    const contentPack = ContentPack.fromJSON(this.props.contentPack);

    return (
      <>
        <Input
          name="comment"
          id="comment"
          type="text"
          maxLength={512}
          value={this.state.comment}
          onChange={this._getComment}
          label="Install Comment"
          help="This comment will be stored with the content pack installation and can be used to describe the reason for this installation."
        />
        {parameterInputs.length > 0 && (
          <>
            <strong>Configure Parameter(s):</strong>
            {parameterInputs}
          </>
        )}
        <EntityCreateShareFormGroup
          description="Search for a User or Team to add as collaborator on entities of this content pack."
          entityType="content pack"
          entityTitle=""
          onSetEntityShare={this._setShareRequest}
        />
        <EntitiesListContainer>
          <ContentPackEntitiesList contentPack={contentPack} readOnly />
        </EntitiesListContainer>
      </>
    );
  }
}

export default ContentPackInstall;
