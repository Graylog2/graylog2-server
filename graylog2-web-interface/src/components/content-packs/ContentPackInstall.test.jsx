import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import 'helpers/mocking/react-dom_mock';

import ContentPack from 'logic/content-packs/ContentPack';
import ContentPackInstall from 'components/content-packs/ContentPackInstall';

describe('<ContentPackInstall />', () => {
  const parameter = {
    title: 'A parameter',
    name: 'PARAM',
    type: 'string',
    default_value: 'parameter',
  };

  const entity = {
    type: {
      name: 'grok_pattern',
      version: '1',
    },
    data: {
      title: { '@type': 'string', '@value': 'franz' },
      descr: { '@type': 'string', '@value': 'hans' },
    },
  };

  const contentPack = ContentPack.builder()
    .id(1)
    .rev(2)
    .name('UFW Grok Patterns')
    .description('Grok Patterns to extract informations from UFW logfiles')
    .summary('This is a summary')
    .url('www.graylog.com')
    .vendor('graylog.com')
    .parameters([parameter])
    .entities([entity])
    .build();

  it('should render a install', () => {
    const wrapper = renderer.create(<ContentPackInstall contentPack={contentPack} />);
    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  it('should call install when called', () => {
    const installFn = jest.fn((id, rev, param) => {
      expect(id).toBe(1);
      expect(rev).toBe(2);
      expect(param).toEqual({ comment: 'Test', parameters: { PARAM: { '@type': 'string', '@value': 'parameter' } } });
    });

    const wrapper = mount(<ContentPackInstall contentPack={contentPack} onInstall={installFn} />);
    wrapper.find('input#comment').simulate('change', { target: { value: 'Test' } });
    wrapper.instance().onInstall();
    expect(installFn.mock.calls.length).toBe(1);
  });

  it('should not call install when parameter is missing', () => {
    const installFn = jest.fn();

    const wrapper = mount(<ContentPackInstall contentPack={contentPack} onInstall={installFn} />);
    wrapper.find('input').at(1).simulate('change', { target: { value: '' } });
    wrapper.instance().onInstall();
    expect(installFn.mock.calls.length).toBe(0);
  });
});
