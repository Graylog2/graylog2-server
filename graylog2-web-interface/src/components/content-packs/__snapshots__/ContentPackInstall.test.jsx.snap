// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ContentPackInstall /> should render a install 1`] = `
<div>
  <div
    className="row"
  >
    <div
      className="col-sm-10 col-sm-offset-1"
    >
      <h2>
        Install comment
      </h2>
      <br />
      <br />
      <div
        className="form-group"
      >
        <label
          className="col-sm-3 control-label"
          htmlFor="comment"
        >
          Comment
        </label>
        <div
          className="col-sm-7"
        >
          <input
            className="form-control"
            id="comment"
            label="Comment"
            maxLength={512}
            name="comment"
            onChange={[Function]}
            placeholder=""
            type="text"
            value=""
          />
          
        </div>
      </div>
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="col-sm-10 col-sm-offset-1"
    >
      <h2>
        Configure Parameter
      </h2>
      <br />
      <br />
      <div
        className="form-group"
      >
        <label
          className="col-sm-3 control-label"
          htmlFor="PARAM"
        >
          A parameter
        </label>
        <div
          className="col-sm-7"
        >
          <input
            className="form-control"
            id="PARAM"
            label="A parameter"
            maxLength={250}
            name="PARAM"
            onChange={[Function]}
            placeholder=""
            required={true}
            type="text"
            value="parameter"
          />
          
        </div>
      </div>
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="col-sm-10 col-sm-offset-1"
    >
      <div>
        <h2>
          Entity list
        </h2>
        <br />
        <div
          className="search"
          style={
            Object {
              "marginTop": 15,
            }
          }
        >
          <form
            className="form-inline"
            onSubmit={[Function]}
          >
            <div
              className="form-group has-feedback"
            >
              <input
                autoComplete="off"
                className="query form-control"
                id="common-search-form-query-input"
                onChange={[Function]}
                placeholder="Enter search query..."
                spellCheck="false"
                style={
                  Object {
                    "width": "auto",
                  }
                }
                type="text"
                value=""
              />
            </div>
            <div
              className="form-group"
              style={
                Object {
                  "marginLeft": 5,
                }
              }
            >
              <button
                className="submit-button btn btn-default"
                disabled={false}
                type="submit"
              >
                Filter
              </button>
            </div>
            <div
              className="form-group"
              style={
                Object {
                  "marginLeft": 5,
                }
              }
            >
              <button
                className="reset-button btn btn-default"
                disabled={false}
                onClick={[Function]}
                type="reset"
              >
                Reset
              </button>
            </div>
          </form>
        </div>
        <div>
          <div
            className="row "
          >
            <div
              className="col-md-12"
            >
              <div
                className="data-table table-responsive"
                id="entity-list"
              >
                <table
                  className="table scrollable"
                >
                  <thead>
                    <tr>
                      <th>
                        Title
                      </th>
                      <th>
                        Type
                      </th>
                      <th>
                        Description
                      </th>
                      <th>
                        Action
                      </th>
                    </tr>
                  </thead>
                  <tbody>
                    <tr>
                      <td
                        className="bigColumns"
                      >
                        franz
                      </td>
                      <td>
                        grok_pattern
                      </td>
                      <td
                        className="bigColumns"
                      >
                        
                      </td>
                      <td>
                        <div
                          className="btn-toolbar"
                          role="toolbar"
                        >
                          <button
                            className="btn btn-xs btn-info"
                            disabled={false}
                            onClick={[Function]}
                            type="button"
                          >
                            Show
                          </button>
                        </div>
                      </td>
                    </tr>
                  </tbody>
                </table>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
